### PR TITLE
Approve mintable

### PR DESCRIPTION
**Pull request by nitinmittal23**

### CODE REVIEW

File: src/root/POSRootChainManager.ts [hunk 1]
```diff
@@ -142,11 +147,16 @@ export default class POSRootChainManager extends ContractsBase {
   }
 
   async approveMaxERC20(rootToken: address, options?: SendOptions) {
-    if (!this.erc20Predicate) {
+    const tokenType = this.posRootChainManager.methods.tokenToType(rootToken).call()
+    if (!tokenType) {
+      throw new Error('tokenType not found. Set it while mapping contract.')
+    }
+    const predicate = this.posRootChainManager.methods.typeToPredicate(tokenType).call()
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:154]: Can you add in one function since we are repeating it multiple times?

**Review by jdkanani — CHANGES_REQUESTED** [review]

File: src/root/POSRootChainManager.ts [hunk 2]
```diff
@@ -44,6 +44,15 @@ export default class POSRootChainManager extends ContractsBase {
       options.posMintableERC1155Predicate || options.network.Main.POSContracts.MintableERC1155PredicateProxy
   }
 
+  async getPredicateAddress(rootToken: address) {
+    const tokenType = this.posRootChainManager.methods.tokenToType(rootToken).call()
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:48]:
```suggestion
    const tokenType = await this.posRootChainManager.methods.tokenToType(rootToken).call()
```

File: src/root/POSRootChainManager.ts [hunk 3]
```diff
@@ -127,11 +136,9 @@ export default class POSRootChainManager extends ContractsBase {
   }
 
   async approveERC20(rootToken: address, amount: BN | string, options?: SendOptions) {
-    if (!this.erc20Predicate) {
-      throw new Error('posERC20Predicate address not found. Set it while constructing MaticPOSClient.')
-    }
+    const predicate = this.getPredicateAddress(rootToken)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:139]:
```suggestion
    const predicate = await this.getPredicateAddress(rootToken)
```

File: src/root/POSRootChainManager.ts [hunk 4]
```diff
@@ -44,6 +44,15 @@ export default class POSRootChainManager extends ContractsBase {
       options.posMintableERC1155Predicate || options.network.Main.POSContracts.MintableERC1155PredicateProxy
   }
 
+  async getPredicateAddress(rootToken: address) {
+    const tokenType = this.posRootChainManager.methods.tokenToType(rootToken).call()
+    if (!tokenType) {
+      throw new Error('Invalid Token Type')
+    }
+    const predicateAddress = this.posRootChainManager.methods.typeToPredicate(tokenType).call()
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:52]:
```suggestion
    const predicateAddress = await this.posRootChainManager.methods.typeToPredicate(tokenType).call()
```

File: src/root/POSRootChainManager.ts [hunk 5]
```diff
@@ -314,13 +315,8 @@ export default class POSRootChainManager extends ContractsBase {
...
 
   async approveERC1155(rootToken: address, options?: SendOptions) {
-    if (!this.erc1155Predicate) {
-      throw new Error('posERC1155Predicate address not found. Set it while constructing MaticPOSClient.')
-    }
-    const txObject = this.getPOSERC1155TokenContract(rootToken, true).methods.setApprovalForAll(
-      this.erc1155Predicate,
-      true
-    )
+    const predicate = this.getPredicateAddress(rootToken)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:318]:
```suggestion
    const predicate = await this.getPredicateAddress(rootToken)
```

File: src/root/POSRootChainManager.ts [hunk 6]
```diff
@@ -239,8 +239,9 @@ export default class POSRootChainManager extends ContractsBase {
     if (options && !token) {
       throw new Error('token address is missing')
     }
+    const predicate = this.getPredicateAddress(token)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:242]:
```suggestion
    const predicate = await this.getPredicateAddress(token)
```

File: src/root/POSRootChainManager.ts [hunk 7]
```diff
@@ -213,21 +217,17 @@ export default class POSRootChainManager extends ContractsBase {
...
 
   async approveAllERC721(rootToken: address, options?: SendOptions) {
-    if (!this.erc721Predicate) {
-      throw new Error('posERC721Predicate address not found. Set it while constructing MaticPOSClient.')
-    }
-    const txObject = this.getPOSERC721TokenContract(rootToken, true).methods.setApprovalForAll(
-      this.erc721Predicate,
-      true
-    )
+    const predicate = this.getPredicateAddress(rootToken)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:229]:
```suggestion
    const predicate = await this.getPredicateAddress(rootToken)
```

File: src/root/POSRootChainManager.ts [hunk 8]
```diff
@@ -213,21 +217,17 @@ export default class POSRootChainManager extends ContractsBase {
     if (options && !token) {
       throw new Error('token address is missing')
     }
+    const predicate = this.getPredicateAddress(token)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:220]:
```suggestion
    const predicate = await this.getPredicateAddress(token)
```

File: src/root/POSRootChainManager.ts [hunk 9]
```diff
@@ -195,11 +201,9 @@ export default class POSRootChainManager extends ContractsBase {
   }
 
   async approveERC721(rootToken: address, tokenId: BN | string, options?: SendOptions) {
-    if (!this.erc721Predicate) {
-      throw new Error('posERC721Predicate address not found. Set it while constructing MaticPOSClient.')
-    }
+    const predicate = this.getPredicateAddress(rootToken)
```
Comment (jdkanani) [inline src/root/POSRootChainManager.ts:204]:
```suggestion
    const predicate = await this.getPredicateAddress(rootToken)
```

**Review by jdkanani — CHANGES_REQUESTED** [review]

**Review by rahuldamodar94 — APPROVED** [review]

**Review by jdkanani — APPROVED** [review]

**Pull request merged**